packages/code-snippet/src/CodeSnippetWidget.tsx

```diff
@@ -214,37 +52,164 @@ export class CodeSnippetWidget extends ReactWidget {
+    this.openCodeSnippetEditor = this.openCodeSnippetEditor.bind(this);
+    this.saveSnippet = this.saveSnippet.bind(this);
+    this.editorFactory = editorFactory;
+    this.app = app;
+    this.app.commands.addCommand(METADATA_EDITOR_ID + ':open', {
+      isVisible: () => {
+        return false;
+      },
+      execute: (args: any) => {
+        this.openCodeSnippetEditor(
+          args.name,
+          args.description,
+          args.language,
+          args.code,
+          args.newFile
+        );
+      }
+    });
+    this.restorer = restorer;
+    this.editorTracker = new WidgetTracker<MetadataEditor>({
+      namespace: METADATA_EDITOR_ID
+    });
+    this.restorer.restore(this.editorTracker, {
+      name: (metadataEditor: any) => {
+        return metadataEditor.metadata.name;
+      },
+      command: METADATA_EDITOR_ID + ':open',
+      args: (widget: MetadataEditor) => ({
+        name: widget.metadata.name,
+        description: widget.metadata.description,
+        language: widget.metadata.language,
+        code: widget.metadata.code,
+        newFile: widget.newFile
+      })
+    });
+    this.fetchData = this.fetchData.bind(this);
+    this.updateSnippets = this.updateSnippets.bind(this);
   }
 
   // Request code snippets from server
   async fetchData(): Promise<ICodeSnippet[]> {
     return await this.codeSnippetManager.findAll();
   }
 
-  // Triggered when the widget button on side palette is clicked
-  onAfterShow(msg: Message): void {
+  updateSnippets(): void {
     this.fetchData().then((codeSnippets: ICodeSnippet[]) => {
       this.renderCodeSnippetsSignal.emit(codeSnippets);
     });
   }
 
+  // Triggered when the widget button on side panel is clicked
+  onAfterShow(msg: Message): void {
+    this.updateSnippets();
+  }
+
+  addCodeSnippet(): void {
+    this.openCodeSnippetEditor('', '', '', '', true);
+  }
+
+  saveSnippet(snippetEditor: MetadataEditor): void {
+    const newSnippet = {
+      schema_name: 'code-snippet',
+      name: snippetEditor.metadata.name,
+      display_name: snippetEditor.metadata.name,
+      metadata: {
+        description: snippetEditor.metadata.description,
+        language: snippetEditor.metadata.language,
+        code: snippetEditor.editor.model.value.text.split('\n')
+      }
+    };
+    const newSnippetString = JSON.stringify(newSnippet);
+
+    if (snippetEditor.newFile) {
+      RequestHandler.makePostRequest(
+        snippetEditor.endpoint,
+        JSON.stringify(newSnippet),
+        false
+      ).then((response: any): void => {
+        this.updateSnippets();
+      });
+      snippetEditor.newFile = false;
+    } else {
+      RequestHandler.makeServerRequest(
+        snippetEditor.endpoint + '/' + newSnippet.name,
+        { method: 'PUT', body: newSnippetString },
+        false
+      ).then((response: any): void => {
+        this.updateSnippets();
+      });
+    }
+  }
+
+  async openCodeSnippetEditor(
```

Comment:
just something to think about...
rather than have callers pass in each property individually could this accept an object containing appropriate fields that you would deconstruct:
```javascript
  async openCodeSnippetEditor({name, description, language, code, newFile=true})
```
or if empty string is preferred default value instead of `undefined`:
```javascript
  async openCodeSnippetEditor({name='', description='', language='', code='', newFile=true})
```

that way each time `openCodeSnippetEditor` is called the caller does not have to worry about provided every single field individually and in the correct order.
for example, in line 79 you could just do something like:
```javascript
this.openCodeSnippetEditor(args)
```

and in line 126 you could do this:
```javascript
this.openCodeSnippetEditor({});
```